src/Microsoft.OData.Core/Json/IStreamBasedJsonWriterFactory.cs

```diff
@@ -0,0 +1,29 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="IStreamBasedJsonWriterFactory.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System;
+using System.IO;
+using System.Text;
+
+namespace Microsoft.OData.Json
+{
+    /// <summary>
+    /// The interface of factories that create <see cref="IJsonWriter"/> instances
+    /// that write directly to a <see cref="Stream"/> rather than a <see cref="TextWriter"/>.
+    /// </summary>
+    [CLSCompliant(false)]
+    public interface IStreamBasedJsonWriterFactory
+    {
+        /// <summary>
+        /// Creates a new JSON writer of <see cref="IJsonWriter"/>.
+        /// </summary>
+        /// <param name="stream">Output stream to which the resulting <see cref="IJsonWriter"/> should write data.</param>
+        /// <param name="isIeee754Compatible">True if it is IEEE754Compatible.</param>
+        /// <param name="encoding">The text encoding of the output data.</param>
+        /// <returns>The JSON writer created.</returns>
+        IJsonWriter CreateJsonWriter(Stream stream, bool isIeee754Compatible, Encoding encoding);

```

Comment:
We really need to divorce this interface from the default implementation that we are providing. If I'm a customer providing my custom implementation, I don't know whether to return `null` or to throw an exception if I can't handle something. We are asserting that the returned instance is not `null`, but we should *immediately throw* in that case based on how the code is currently written. We assume that it's not `null`, and we fail catastrophically later on if it is `null`, so let's just throw at the moment the issue occurs. *This was my entire point in the other thread*. Someone who implements this will do something silly, and we need to be able to help them figure out the problem immediately. If `CreateJsonWriter` throws, we should wrap that exception in something that says "you provided a customer jsonwriter factory and it failed for these input parameters". Further, if it returns `null`, we should throw an exception saying the same. The fact that our default writer doesn't run into these situations is inconsequential.

Reply:
@corranrogue9 I did already make the change to throw when the `IJsonStreamBasedJsonWriterFactory` returns null. What I can do in addition, if that is not sufficient,  is include details about the input parameters in the error message.

However, I did not change the behaviour in relation to the existing interface `IJsonWriterFactory`. It currently does not throw an exception when it returns null. I left that as-is. We have avoided implemented things are almost objective improvements to existing behaviour because they might result in breaking changes in some edge cases. Throwing an exception when the return value is null in this case may also be a breaking changes if the customer happens to depend on it not throwing an exception or if the customer has code that explicitly handles a `NullReferenceException`. 

I would not be comfortable changing this behaviour unless we explcitly acknowledge that this could be a breaking change (like we have done in other scenarios) and agree that we are okay with such outcomes, especially given that:
- The code checks for null via a `Debug.Assert` statement, meaning that the previous authors acknowledged that the factory could return null, but opted not to handle it as an exception for some reason.
- No one has complained about this yet or reported an issue related to this. I concede that this is probably because it's highly unlikely people are creating their own customer `IJsonWriter` implementations.

Reply:
I think I got lost somewhere, I didn't mean you should change the `IJsonWriterFactory` handling, that seems out of scope for this PR, thanks for the explanation though. Thanks for adding the logic for `IStreamBasedJsonWriterFactory` returning `null`. 

I think this kind of "breaking change" would be a good discussion topic though :) At some point, a caller can depend on basically any behavior (including non-functional requirements like performance issues), so we have to draw the line somewhere.